plugins/modules/foreman_compute_resource.py

```diff
@@ -112,133 +154,93 @@
-        sanitize_entity_dict,
-    )
-
-    from nailgun.entities import (
-        AbstractComputeResource,
-        LibvirtComputeResource,
-        OVirtComputeResource,
-        VMWareComputeResource,
-    )
-
-    has_import_error = False
-except ImportError as e:
-    has_import_error = True
-    import_error_msg = str(e)
-
-from ansible.module_utils.foreman_helper import ForemanEntityAnsibleModule
-
-# This is the only true source for names (and conversions thereof)
-name_map = {
-    'name': 'name',
-    'description': 'description',
-    'provider': 'provider',
-    'organizations': 'organization',
-    'locations': 'location',
-    'datacenter': 'datacenter',
-}
+from ansible.module_utils.foreman_helper import ForemanEntityApypieAnsibleModule
 
 
 def get_provider_infos(provider):
     provider_name = provider.lower()
 
     if provider_name == 'libvirt':
         return {
-            'params': ['url', 'display_type'],
-            'class': LibvirtComputeResource
+            'params': ['url', 'display_type']
         }
 
     elif provider_name == 'ovirt':
         return {
-            'params': ['url', 'user', 'password', 'datacenter', 'use_v4', 'ovirt_quota'],
-            'class': OVirtComputeResource
+            'params': ['url', 'user', 'password', 'datacenter', 'use_v4', 'ovirt_quota']
         }
 
     elif provider_name == 'vmware':
         return {
-            'params': ['url', 'user', 'password', 'datacenter'],
-            'class': VMWareComputeResource
+            'params': ['url', 'user', 'password', 'datacenter']
         }
 
     else:
         return {
-            'params': [],
-            'class': AbstractComputeResource
+            'params': []
         }
 
 
 def main():
-    module = ForemanEntityAnsibleModule(
-        argument_spec=dict(
-            name=dict(type='str', required=True),
-            updated_name=dict(type='str'),
-            description=dict(type='str'),
-            provider=dict(type='str', choices=['vmware', 'libvirt', 'ovirt']),
+    module = ForemanEntityApypieAnsibleModule(
+        entity_spec=dict(
+            name=dict(required=True),
+            updated_name=dict(),
+            description=dict(),
+            organizations=dict(type='entity_list', flat_name='organization_ids'),
+            locations=dict(type='entity_list', flat_name='location_ids'),
+            provider=dict(choices=['vmware', 'libvirt', 'ovirt']),
             provider_params=dict(type='dict'),
-            locations=dict(type='list'),
-            organizations=dict(type='list'),
-            state=dict(type='str', default='present', choices=['present', 'absent', 'present_with_defaults']),
-
-            # Deprecated provider's specific params, use nested keys in provider_params param instead
-            provider_auth=dict(type='dict'),
-            url=dict(type='str'),
-            display_type=dict(type='str'),
-            datacenter=dict(type='str'),
+            display_type=dict(),
```

Comment:
shouldn't all these be nested  under `provider_params` or would that require #385 for the nesting to work?

Reply:
No, this sort of nesting works in the old way.